Add code for argument information processing

Aijuh opened the pull request:
This closes issue #238. 

Main part of this PR is to add code for processing argument information at _ScanEntry class.

## PR description
Scannable argInfo has following information.

![image](https://github.com/snu-quiqcl/iquip/assets/43592775/9fdcb43c-0037-445f-b3bf-0a5f2d99b478)

`default` is an dictionary that contains argument of each scan type. If scan type differs, its key and value are different. This information is called as `state` in this PR.
`unit`, `scale`, and etc are common to all scan type. This information is called as `procdesc` in this PR.
This information is used when widget for each scan type is created. 

For this PR, I did not implemented widget for each scan type. But this widget will use `state` and `procdesc` at the following PRs.

## Example image
![image](https://github.com/snu-quiqcl/iquip/assets/43592775/00a9996c-86e8-4f79-ac8c-b3c55a42ff8e)

Aijuh committed:
ArgInfo processing & empty class

BECATRUE commented:
Oh, this PR is also too long.
Could you separate it into several PRs?

Aijuh committed:
Parameter process

Aijuh commented:
I deleted all the scan type classes.
I will add RangeScan type first at the next PR.

Aijuh committed:
Delete unused method

Aijuh committed:
TODO comment

BECATRUE commented on iquip/apps/builder.py:
```diff
@@ -309,11 +312,83 @@ def __init__(
             name: See the attributes section.
-            argInfo: See the attributes section.
+            argInfo: Each key and its value are:
+              default: The dictionary that describes arguments of a scannable object.
```
I think "the default scannable object" is better instead of "a scannable object", because it describes "default" options.

Aijuh replied:
You are right.

BECATRUE commented on iquip/apps/builder.py:
```diff
@@ -309,11 +312,83 @@ def __init__(
+              global_step: The step between values changed by the up and down button.
+              global_min: The minimum value. (default=float("-inf"))
+              global_max: The maximum value. (default=float("inf"))
+                If min > max, then they are swapped.
```
Is this supported by ARTIQ, or will you implement this function?

Aijuh replied:
This min, max value will be used in QSpinBox, which has following properties ([ref](https://doc.qt.io/qt-6/qspinbox.html#minimum-prop)).
They will be adjusted automatically.

BECATRUE commented on iquip/apps/builder.py:
```diff
@@ -309,11 +312,83 @@ def __init__(
+        self.layout.addWidget(self.stack, 1, 1)
+
+    def get_state(self, argInfo):
+        """Gets a state dictionary that describes key parameters of a scannable object.
```
How about "the default scannable object"?

BECATRUE commented on iquip/apps/builder.py:
```diff
@@ -309,11 +312,83 @@ def __init__(
+        Creates a default state dictionary and updates it using the argInfo.
+
+        Args:
+            argInfo: A dictionary that contains argument information of each scannable type.
```
Since you already wrote the description of `argInfo`, please write "argInfo: See argInfo in __ init __()." instead.

BECATRUE replied:
Oh, please remove two blanks before and after "init".
I added those to prevent it from turning into bold according to markdown syntax.

BECATRUE commented on iquip/apps/builder.py:
```diff
@@ -309,11 +312,83 @@ def __init__(
+    def get_state(self, argInfo):
+        """Gets a state dictionary that describes key parameters of a scannable object.
+
+        Creates a default state dictionary and updates it using the argInfo.
```
It looks unnecessary as the first line is sufficient.

BECATRUE commented on iquip/apps/builder.py:
```diff
@@ -309,11 +312,83 @@ def __init__(
+        self.layout.addWidget(QLabel(name, self), 0, 0)
+        self.layout.addWidget(self.stack, 1, 1)
+
+    def get_state(self, argInfo):
```
Please add the type hinting in terms of `argInfo` and return value.

BECATRUE commented on iquip/apps/builder.py:
```diff
@@ -309,11 +312,83 @@ def __init__(
 
```
I realized this now. Please remove one tab before `):` like below.
```python
def __init__(
  self,
  name: str,
  argInfo: Dict[str, Any],
  parent: Optional[QWidget] = None
):
```

BECATRUE commented on iquip/apps/builder.py:
```diff
@@ -291,13 +291,16 @@ def value(self) -> Optional[str]:
     Attributes:
         name: The name of the scannable object.
-        argInfo: The infomation of the arguments.
+        state: The dictionary that describes argument information of each scan type.
```
I think you should write more description about the structure of `state`.
However, if you improve it in next PR, it is okay to remain it in this PR.

Aijuh replied:
I changed docstring as below. 
Detailed info of dictonary for each scannble type will be described in next PR. They will be added at each scannable type Entry class.

```
        state: Each key and its value are:
          "selected": The name of the selected scannable type.
          "NoScan": The dictionary that contains argument info of NoScan scannable type.
          "RangeScan": The dictionary that contains argument info of RangeScan scannable type.
          "CenterScan": The dictionary that contains argument info of CenterScan scannable type.
          "ExplicitScan": The dictionary that contains argument info of ExplicitScan scannable type.
```

BECATRUE commented on iquip/apps/builder.py:
```diff
@@ -291,13 +291,16 @@ def value(self) -> Optional[str]:
 
-# TODO(AIJUH): Add feature for argInfo processing and other scan type classes.
+# TODO(AIJUH): Add other scan type classes.
 class _ScanEntry(QWidget):
```
I have a question. Why did you inherit from `QWidget` directly, not `_BaseEntry`?

Aijuh replied:
`_BaseEntry` has value method, which returns selected value. When I first thought of it, I didn't inherit from `_BaseEntry` because `_ScanEntry` doesn't return a single value.

Moreover, since `_ScanEntry` use QGridLayout for design, I did not start from `_BaseEntry`.

BECATRUE replied:
I see. There seems to be several differences between `_BaseEntry` and `_ScanEntry`.

BECATRUE commented on iquip/apps/builder.py:
```diff
@@ -291,13 +291,16 @@ def value(self) -> Optional[str]:
         name: The name of the scannable object.
-        argInfo: The infomation of the arguments.
+        state: The dictionary that describes argument information of each scan type.
+        stackWidget: The QstackWidget that contains widget of each scan type.
```
"widgets"

BECATRUE commented on iquip/apps/builder.py:
```diff
@@ -309,11 +312,83 @@ def __init__(
-        self.argInfo = argInfo
+        self.state = self.get_state(argInfo)
+        self.stack = QStackedWidget(self)
+        self.layout = QGridLayout(self)
```
Why did you use `QGridLayout`?
Could you explain your plan briefly?

Aijuh replied:
I followed original artiq code in terms of design.

BECATRUE commented on iquip/apps/builder.py:
```diff
@@ -309,11 +312,83 @@ def __init__(
+        """Gets a prodesc dictionary that describes common parameters of the scannable object.
+
+        Args:
+            argInfo: A dictionary that contains argument information of the scannable object.
```
Please update `argInfo` docstring like the review above.

BECATRUE replied:
Please update to `__init__()`.

BECATRUE commented on iquip/apps/builder.py:
```diff
@@ -309,11 +312,83 @@ def __init__(
+        return state
+
+    def get_procdesc(self, argInfo):
+        """Gets a prodesc dictionary that describes common parameters of the scannable object.
```
"procdesc"

BECATRUE commented on iquip/apps/builder.py:
```diff
@@ -309,11 +312,83 @@ def __init__(
+                    logger.warning("unknown default type: %s", ty)
+        return state
+
+    def get_procdesc(self, argInfo):
```
Please write type hintings.

BECATRUE commented on iquip/apps/builder.py:
```diff
@@ -309,11 +312,83 @@ def __init__(
+            "unit": argInfo["unit"],
+            "scale": argInfo["scale"],
+            "global_step": argInfo["global_step"],
+            "ndecimals": argInfo["ndecimals"],
```
In according to your docstring, please move `ndecimals` to the last.

BECATRUE commented on iquip/apps/builder.py:
```diff
@@ -309,11 +312,83 @@ def __init__(
+            "global_min": argInfo["global_min"],
+            "global_max":  argInfo["global_max"]
+        }
+        return procdesc
```
I think you can return this dict directly, not defining a variable.
However it is just a coding style, do as you please.

Aijuh replied:
I will reconsider it at the next PR.

BECATRUE commented on iquip/apps/builder.py:
```diff
@@ -309,11 +312,83 @@ def __init__(
+        """
+        scale = argInfo["scale"]
+        state = {
+            "selected": "NoScan",
```
Is it a temporary setting?

Aijuh replied:
corret

BECATRUE commented on iquip/apps/builder.py:
```diff
@@ -309,11 +312,83 @@ def __init__(
+        state = {
+            "selected": "NoScan",
+            "NoScan": {"value": 0.0, "repetitions": 1},
+            "RangeScan": {"start": 0.0, "stop": 100.0*scale, "npoints": 10,
```
Please add a space before and after an operator like `100.0 * scale`.

BECATRUE commented on iquip/apps/builder.py:
```diff
@@ -309,11 +312,83 @@ def __init__(
+            "NoScan": {"value": 0.0, "repetitions": 1},
+            "RangeScan": {"start": 0.0, "stop": 100.0*scale, "npoints": 10,
+                          "randomize": False, "seed": None},
+            "CenterScan": {"center": 0.*scale, "span": 100.*scale,
```
Please replace "span" with "step" same as #111.

Aijuh replied:
Do you want me to change variable name?
What do you want to say with this #111 reference?

Aijuh replied:
This PR has too many content.
Can you explain me in more detail?

BECATRUE replied:
Oh, I'm really sorry that I confused you.

I intended to say snu-quiqcl/artiq-proxy#111.
I want just to replace the positions of "span" and "step" as follows:
```python
{"center": ..., "step": ..., "span": ..., ...}
```

BECATRUE commented on iquip/apps/builder.py:
```diff
@@ -309,11 +312,83 @@ def __init__(
+                           "seed": None},
+            "ExplicitScan": {"sequence": []}
+        }
+        if "default" in argInfo:
```
Is it possible that the default is not given?

Aijuh replied:
With reference to [this](https://github.com/snu-quiqcl/artiq-proxy/discussions/111), scan object always has default.
Does argInfo for other non scan object has default?

Aijuh replied:
As I checked, it seems that argInfo always has `default`

BECATRUE replied:
Then, did you leave it in case you didn't know?

BECATRUE commented on iquip/apps/builder.py:
```diff
@@ -309,11 +312,83 @@ def __init__(
+        }
+        if "default" in argInfo:
+            defaults = argInfo["default"]
+            if not isinstance(defaults, list):
```
When is its type not a list?

Aijuh replied:
As I tested, `default` is a list type that contains dictionary. 
But there might be the case that I don't know, so I followed original artiq code.

BECATRUE commented on iquip/apps/builder.py:
```diff
@@ -309,11 +312,83 @@ def __init__(
+                elif ty == "ExplicitScan":
+                    state[ty]["sequence"] = default["sequence"]
+                else:
+                    logger.warning("unknown default type: %s", ty)
```
Please start with an upper case.

BECATRUE replied:
And I think "scan type" is better instead of "default type".

BECATRUE commented on iquip/apps/builder.py:
```diff
@@ -309,11 +312,83 @@ def __init__(
+            if not isinstance(defaults, list):
+                defaults = [defaults]
+            state["selected"] = defaults[0]["ty"]
+            for default in reversed(defaults):
```
Why did you reverse `defaults`?

Also, I think this for loop can be shorten as follows:
```python
for default in defaults:
  ty = default.pop("ty")
  if ty not in ["NoScan", "RangeScan", "CenterScan", "ExplicitScan"]:
    logger.warning("Unknown scan type: %s", ty)
  state[ty] = default
```

Aijuh replied:
I wanted to make sure that each dictionary for each scan type in `state` contains only designated keys.
But since each scannable entry will access this dictionary with specific key, this implementation seems good.

Aijuh replied:
I just used artiq code for reversing, but I also think it is meaningless.

Aijuh committed:
docstring update

Aijuh commented:
Thank you for the review @BECATRUE. Even if I double checked docstrings and fixed my codes, there were many things to improve. I changed all the point you gave me.

In terms of using layout, using QGridLayout is necessary for the design.
![image](https://github.com/snu-quiqcl/iquip/assets/43592775/c58f916d-4a87-498d-8e4e-d73bb4f264bc)
In the example image above, button box occupy the upper row. And QLabel and QSpinbox share one row. This design can be done by using QGridLayout. So I followed original artiq code in terms of design.

BECATRUE commented:
I think you didn't apply all reviews such as type hintings or docstrings.
Please check it again!

BECATRUE commented on iquip/apps/builder.py:
```diff
@@ -297,8 +297,13 @@ class _ScanEntry(QWidget):
         name: The name of the scannable object.
-        state: The dictionary that describes argument information of each scan type.
-        stackWidget: The QstackWidget that contains widget of each scan type.
+        state: Each key and its value are:
```
I think there are too many columns, hence how about "Each key and its value are as follows."?

BECATRUE commented on iquip/apps/builder.py:
```diff
@@ -354,39 +357,27 @@ def get_state(self, argInfo):
-            "ndecimals": argInfo["ndecimals"],
             "global_min": argInfo["global_min"],
-            "global_max":  argInfo["global_max"]
+            "global_max":  argInfo["global_max"],
```
Please remove one blank.

BECATRUE reviewed (changes requested):
Both "scannable" and "scan" are used, so how about unifying them into one?

Aijuh commented:
I will follow https://github.com/snu-quiqcl/artiq-proxy/discussions/111 for the order of variable in later PR's docstring.

```
                   "seed": None},
    "ExplicitScan": {"sequence": []}
}
if "default" in argInfo:

```
For this code, I just leave it because there might be the case that default is not given. If this case happens, we can select scan type and its parameters at our GUI.

Aijuh committed:
docstring fixed

BECATRUE commented on iquip/apps/builder.py:
```diff
@@ -291,29 +291,95 @@ def value(self) -> Optional[str]:
+            "NoScan": {"value": 0.0, "repetitions": 1},
+            "RangeScan": {"start": 0.0, "stop": 100.0 * scale, "npoints": 10,
+                          "randomize": False, "seed": None},
+            "CenterScan": {"center": 0. * scale,"step": 10. * scale,
```
Please add a space before "step".

BECATRUE reviewed (changes requested):
Final!

Aijuh committed:
Format change

BECATRUE reviewed (approved):
LGTM!
